core/net/rpl/rpl-icmp6.c

```diff
@@ -721,84 +783,121 @@ dao_input(void)
       }
-      if(flags & RPL_DAO_K_FLAG) {
-        dao_ack_output(instance, &dao_sender_addr, sequence);
-      }
+    } 
+    /* independent if we remove or not - ACK the request */
+    if(flags & RPL_DAO_K_FLAG) {
+      /* indicate that we accepted the no-path DAO */
+      dao_ack_output(instance, &dao_sender_addr, sequence,
+                     RPL_DAO_ACK_UNCONDITIONAL_ACCEPT);
     }
     goto discard;
   }
 
   PRINTF("RPL: adding DAO route\n");
 
-  if((nbr = uip_ds6_nbr_lookup(&dao_sender_addr)) == NULL) {
-    if((nbr = uip_ds6_nbr_add(&dao_sender_addr,
-                              (uip_lladdr_t *)packetbuf_addr(PACKETBUF_ADDR_SENDER),
-                              0, NBR_REACHABLE)) != NULL) {
-      /* set reachable timer */
-      stimer_set(&nbr->reachable, UIP_ND6_REACHABLE_TIME / 1000);
-      PRINTF("RPL: Neighbor added to neighbor cache ");
-      PRINT6ADDR(&dao_sender_addr);
-      PRINTF(", ");
-      PRINTLLADDR((uip_lladdr_t *)packetbuf_addr(PACKETBUF_ADDR_SENDER));
-      PRINTF("\n");
-    } else {
-      PRINTF("RPL: Out of Memory, dropping DAO from ");
-      PRINT6ADDR(&dao_sender_addr);
-      PRINTF(", ");
-      PRINTLLADDR((uip_lladdr_t *)packetbuf_addr(PACKETBUF_ADDR_SENDER));
-      PRINTF("\n");
-      goto discard;
+  /* Update and add neighbor - if no room - fail. */
+  if((nbr = rpl_icmp6_update_nbr_table(&dao_sender_addr, NBR_TABLE_REASON_RPL_DAO, instance)) == NULL) {
+    PRINTF("RPL: Out of Memory, dropping DAO from ");
+    PRINT6ADDR(&dao_sender_addr);
+    PRINTF(", ");
+    PRINTLLADDR((uip_lladdr_t *)packetbuf_addr(PACKETBUF_ADDR_SENDER));
+    PRINTF("\n");
+    if(flags & RPL_DAO_K_FLAG) {
+      /* signal the failure to add the node */
+      dao_ack_output(instance, &dao_sender_addr, sequence,
+		     is_root ? RPL_DAO_ACK_UNABLE_TO_ACCEPT_ROOT :
+		     RPL_DAO_ACK_UNABLE_TO_ACCEPT);
     }
-  } else {
-    PRINTF("RPL: Neighbor already in neighbor cache\n");
+    goto discard;
   }
 
   rep = rpl_add_route(dag, &prefix, prefixlen, &dao_sender_addr);
   if(rep == NULL) {
     RPL_STAT(rpl_stats.mem_overflows++);
     PRINTF("RPL: Could not add a route after receiving a DAO\n");
+    if(flags & RPL_DAO_K_FLAG) {
+      /* signal the failure to add the node */
+      dao_ack_output(instance, &dao_sender_addr, sequence,
+		     is_root ? RPL_DAO_ACK_UNABLE_TO_ACCEPT_ROOT :
+		     RPL_DAO_ACK_UNABLE_TO_ACCEPT);
+    }
     goto discard;
   }
 
+  /* set lifetime and clear NOPATH bit */
   rep->state.lifetime = RPL_LIFETIME(instance, lifetime);
-  rep->state.learned_from = learned_from;
-  rep->state.nopath_received = 0;
+  RPL_ROUTE_CLEAR_NOPATH_RECEIVED(rep);
 
 #if RPL_CONF_MULTICAST
 fwd_dao:
 #endif
 
   if(learned_from == RPL_ROUTE_FROM_UNICAST_DAO) {
+    int should_ack = 0;
+
+    if(flags & RPL_DAO_K_FLAG) {
+      /* check if this route is already installed and we can ack now! */
+      /* not pending - and same seq-no means that we can ack.         */
```

Comment:
This means that the route is already installed - e.g. the route have already been installed due to previous DAO / DAO_ACK interactions and therefore we know that it will be fine this time also - it is an optimisation to avoid waiting for the end-to-end ack this time. I will improve the comment a bit.